### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by cbarraford**
I was trying to solve that a url like "http://example.com/foo#bar" doesn't return any hashtags (ie "#bar") and found the source code really hard to follow. Lots of variables that aren't clear what value they hold, lots of indentations, etc.

So I refactored with a different approach that is simpler to follow, significantly less code, and better commented. There are some behavior changes....

the GetTags func now takes a string instead of io.Reader
"@gernest@mwanza" not returns [gernest@mwanza] instead of [gernest, mwanza]
Having your char also be a terminator results in not the behavior you want.
Some of this may be controversial, happy to have that discussion.

### CODE REVIEW

**Review by gernest — APPROVED** [review]
Thanks. This looks nice to me.  When I started this project I was more into generic approach.

It make sense to require a space before the terminator which will be a correct mention behaviour .

**Comment by gernest** [comment]
@cbarraford you can merge this. When I get time I will add benchmarks and go-fuzz this so we can easily detect regressions in the future.

**Comment by cbarraford** [comment]
👍
